drivers/xen/gnttab.c

```diff
@@ -332,13 +328,39 @@ const char *gnttabop_error(int16_t status)
 	}
 }
 
+/* Picked from Linux implementation */
+#define LEGACY_MAX_GNT_FRAMES_SUPPORTED		4
+static unsigned long gnttab_get_max_frames(void)
+{
+	int ret;
+	struct gnttab_query_size q = {
+		.dom = DOMID_SELF,
+	};
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_query_size, &q, 1);
+	if ((ret < 0) || (q.status != GNTST_okay)) {
+		return LEGACY_MAX_GNT_FRAMES_SUPPORTED;
+	}
+
+	return q.max_nr_frames;
+}
+
 static int gnttab_init(void)
 {
 	grant_ref_t gref;
 	struct xen_add_to_physmap xatp;
 	struct gnttab_setup_table setup;
-	xen_pfn_t frames[NR_GRANT_FRAMES];
+	xen_pfn_t frames[CONFIG_NR_GRANT_FRAMES];
 	int rc = 0, i;
+	unsigned long xen_max_grant_frames;
+	uintptr_t gnttab_base = DT_REG_ADDR_BY_IDX(DT_INST(0, xen_xen), 0);
+	mm_reg_t gnttab_reg;
+
+	xen_max_grant_frames = gnttab_get_max_frames();
+	if (xen_max_grant_frames < CONFIG_NR_GRANT_FRAMES) {
+		LOG_ERR("Xen max_grant_frames is less than CONFIG_NR_GRANT_FRAMES!");
```

Comment:
This logic should be reversed, no?
-  if Xen allows more GNT that we can support  it's a problem, no way to serve them. ASSERt will hit in get_free_entry(). no?
- If less  -it's ok and just upper limit need to be adjusted.

or it should be explicit for now (xen_max_grant_frames != CONFIG_NR_GRANT_FRAMES).

Am i missing smth?

Reply:
No, logic in code is correct. When Xen allows you to have more frames than you need - it is completely OK, just do not use and map them.
If you need more, than hypervisor allows - you should panic.

Reply:
ok. noted.